### PR TITLE
EntityMetadata - Implement getOptions method

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
New canonical source for field options. Part of [dev/core#5410](https://lab.civicrm.org/dev/core/-/issues/5410)

This includes a possibly-breaking-but-mostly-backward-compat change to the signature of `hook_civicrm_fieldOptions`, however that hook was previously quite difficult to work with and prone to bugs, and this new signature significantly improves the hook usability & stability.
[Documenation update for hook_civicrm_fieldOptions](https://lab.civicrm.org/documentation/docs/dev/-/merge_requests/1190).

Before
----------------------------------------
Many years ago, @twomice and I sat down at a sprint and created the `CRM_Core_Pseudoconstant::get` function. It's served us well, but also has some shortcomings:

- Returns a flat array - this was fine for a traditional `<select>`, but these days many of our option lists and widgets support more info like descriptions, abbreviations, icons and colors.
- An overloaded `"context"` param would simultaneously control whether disabled options were returned, whether keys and values should be swapped, or would return alternate values like `name` or `abbr`.
- Other params like `"flip"` were more confusing than useful.
- Support for permission checks was spotty.
- All these different contexts and transformations make `hook_civicrm_fieldOptions` very difficult to implement - you'd have to understand all of the implications of each context and take each variant into account or else your hook would be prone to random bugs.

After
----------------------------------------
New `Entity->getOptions()` method is envisioned as the new canonical source for field options, supplanting `BAO::buildOptions` and `CRM_Core_Pseudoconstant::get`.

It always returns a multidimensional array, solving the age-old problem of what keys and values to return.

It calls `hook_civicrm_fieldOptions`, with `'context' = 'full'` to indicate the new format being used. Once this migration is complete, the hook will be far easier to use, as the format will always be consistent. For backward-compat with legacy hook implementations, this function will convert scalar values supplied via hook into the new full format.

Technical Details
------------------
This switches `DAO::buildOptions` to call the new `Entity->getOptions()` method, and deletes most of the old `BAO::buildOptions` overrides, replacing them with either a preprocessing callback (`condition_provider`), or a postprocessing callback (`hook_civicrm_fieldOptions`).
`CRM_Core_Pseudoconstant::get` still fetches options the old way but is marked deprecated & its use is to be phased out.

### CODE REVIEW

File: CRM/Campaign/BAO/Survey.php [hunk 1]
```diff
@@ -949,27 +949,12 @@ public static function releaseRespondent($params) {
...
-    switch ($fieldName) {
-      case 'activity_type_id':
-        $campaignCompId = CRM_Core_Component::getComponentID('CiviCampaign');
-        if ($campaignCompId) {
-          $params['condition'] = ["component_id={$campaignCompId}"];
-        }
-        break;
-    }
-    return CRM_Core_PseudoConstant::get(__CLASS__, $fieldName, $params, $context);
+  public static function alterActivityTypes(string $fieldName, CRM_Utils_SQL_Select $conditions) {
```
Comment (monishdeb) [inline CRM/Campaign/BAO/Survey.php:955]: @colemanw did you miss to add the third fn argument $params here ?

Comment (colemanw) [reply CRM/Campaign/BAO/Survey.php:955]: @monishdeb you're right that is passed to the callback, but in this case it's not used by the function so I just left it off.

File: CRM/Event/BAO/Event.php [hunk 2]
```diff
@@ -2247,30 +2247,15 @@ public static function updateParticipantCampaignID($eventID, $eventCampaignID) {
...
-        // @todo - the component is enabled check should be done within getIncomeFinancialType
-        // It looks to me like test cover was NOT added to cover the change
-        // that added this so we need to assume there is no test cover
-        if (CRM_Core_Component::isEnabled('CiviContribute')) {
-          return CRM_Financial_BAO_FinancialType::getIncomeFinancialType($props['check_permissions'] ?? TRUE);
-        }
-        return [];
-    }
-    return CRM_Core_PseudoConstant::get(__CLASS__, $fieldName, $params, $context);
+  public static function getFinancialType() {
```
Comment (colemanw) [inline CRM/Event/BAO/Event.php:2252]: @monishdeb this looks good but this function is missing params:
```suggestion
  public static function getFinancialType($fieldName, $params) {
```

File: CRM/Event/BAO/Participant.php [hunk 3]
```diff
@@ -1743,35 +1743,30 @@ public static function addActivityForSelection($participantId, $activityType) {
...
       // FIXME: Why does this option even exist if cart is disabled?
       if (!Civi::settings()->get('enable_cart')) {
-        $params['condition'][] = "name <> 'Pending in cart'";
+        $conditions->where('name <> "Pending in cart"');
       }
     }
-    if ($fieldName === 'status_id' && isset($props['is_counted'])) {
-      $params['condition'][] = 'is_counted = ' . $props['is_counted'];
+    if (isset($params['values']['is_counted'])) {
+      $conditions->where('is_counted = #counted', ['counted' => $params['values']['is_counted']]);
```
Comment (colemanw) [inline CRM/Event/BAO/Participant.php:1768]: Probably we ought to cast this to `(int)` just to be safe.

**Comment by monishdeb** [comment]
* General standards
    * ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) __PASS__
    * ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) __PASS__ : A doc MR https://lab.civicrm.org/documentation/docs/dev/-/merge_requests/1190 is submitted for the fieldOptions hook signature changes and is back compatible. The MR description mention a implication what happen if $option's value is treated as string but not array of attributes. 
   * ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) __PASS__: Tested the patch on local setup for each entities whose buildOptions are replaced or deprecated and doesn't break backoffice/front-end forms, APIv3/v4 output. 
    * ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) __PASS__: After reviewing the patch I can confirm that there won't be any affect any functionality for the fact that DAO::buildOptions hasn't been replaced but now internally calls EntityRepository->getOptions() - more robust, flexible for any customization, support both real and pseudo/virtual entities. Although some of the BAO::buildOptions overrides are not replaced : 
 ```bash
 CRM/Contact/BAO/GroupContact.php:749:  public static function buildOptions($fieldName, $context = NULL, $props = []) {
CRM/Contact/BAO/Contact.php:3306:  public static function buildOptions($fieldName, $context = NULL, $props = []) {
CRM/Contact/BAO/Relationship.php:2010:  public static function buildOptions($fieldName, $context = NULL, $props = []) {
CRM/Core/DAO.php:2890:  public static function buildOptions($fieldName, $context = NULL, $values = []) {
CRM/Core/BAO/Address.php:1253:  public static function buildOptions($fieldName, $context = NULL, $props = []) {
CRM/Core/BAO/EntityTag.php:443:  public static function buildOptions($fieldName, $context = NULL, $props = []) {
CRM/Contribute/BAO/ContributionRecur.php:1023:  public static function buildOptions($fieldName, $context = NULL, $props = []) {
CRM/Contribute/BAO/Contribution.php:3296:  public static function buildOptions($fieldName, $context = NULL, $props = []) {
CRM/Case/BAO/Case.php:2977:  public static function buildOptions($fieldName, $context = NULL, $props = []) {
```
as the internal overrides cannot be entirely shifted to condition_provider(s). But then these overrides are marked deprecated and eventually calls DAO::buildOptions >> EntityRepository->getOptions(). The other BAO::buildOptions overrides are replaced by `condition_provider` or `callback` as declaration in schema/*.entityType.php and postprocessing callback via hook_civicrm_fieldOptions. This related PR https://github.com/civicrm/civicrm-core/pull/31021 need mention here that replaces CRM_Core_PseudoConstant::get call with DAO::buildOptions.  

    
* Developer standards
    * ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __PASS__: I think the technical details provide enough gist.  
    * ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __PASS__
    * ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __PASS__
    * ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) __PASS__ : Modification to the existing unit-tests is apt and the new unit-tests rightfully examine the output of APIv4 `Entity::getFields`.

**Comment by totten** [comment]
> ([r-doc](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) PASS : A doc MR https://lab.civicrm.org/documentation/docs/dev/-/merge_requests/1190 is submitted for the fieldOptions hook signature changes and is back compatible. The MR description mention a implication what happen if $option's value is treated as string but not array of attributes.

Woot!

> (___PR___) It calls hook_civicrm_fieldOptions, with 'context' = 'full' to indicate the new format being used.
>
> (___Doc___)  If you need to support older versions of Civi, or legacy subsytems like CiviReport, check the `context` param
and add extra handling for the various possible formats.

OK, when I read this part, I had a bit of a heart palpitation -- changing the format of `array $options` and also changing the interpretation of `$params['context']`  sounds like an invitation for some twisty-turny situations. (From a contract POV, it sounds like enough to consider a different hook name -- with an adapter/deprecation for the old hook name.*)

But then I read `universe` and relaxed. I think I understand the basic rationale here -- i.e. why most extensions would be safe. But these things can persnickity, so I'm just gonna verbalize anyway.

* __Before__: A typical hook implementer uses the hook to append an item (or to assign an entire list), e.g.
    ```php
    function hook_civicrm_fieldOptions($entity, $field, &$options, $params) {
      if ("$entity.$field" === "SomeEntity.some_field") {
        $options[123] = 'Hello world';     // Append one item
        $options = [123 => 'Hello world'];  // Or assign an entire list of items
        $options += [123 => 'Hello world']; // Or append an entire list of items
      }
    }
    ```
* __After__: The above example is still valid. Additionally, when appending/assigning items, you may replace the value with an array.
    ```php
    function hook_civicrm_fieldOptions($entity, $field, &$options, $params) {
      if ("$entity.$field" === "SomeEntity.some_field") {
        
        $options[123] = 'Hello world';         // Append one item
        // $options = [123 => 'Hello world'];  // Or assign an entire list of items
        // $options += [123 => 'Hello world']; // Or append an entire list of items
      }
    }
    ```
* __Analysis__
    * The safety of this comes from the fact that the vast majority of implementations are following the same pattern ("Before") -- and that this pattern still leads to the same.
    * The risk is that extensions could do *other things*. It would be risky to:
        *  Read/scan/filter the inputted values of `$options`.  (They can't expect values to be `string`s -- they should look for `string|array`.)
        * Interpret/adapt based on the `$params['context']`. (If an old `context` is replaced by a new `context`. Conditionals on the old `context` won't fire the same way.)
    * In `universe`, I found ~17 extensions referencing the hook. To my first estimation (*no testing*), 15 extensions seemed to land on the typical/safe side and 2 on the risky/problematic side.
    * If those numbers are right, then I think it's fair

**Comment by totten** [comment]
(*Ooops, last comment was submitted while I was still editing... so the conclusion...*)

I'm OK with the `merge ready`, and I don't want to block. However, my comments are based on some loose-reading/snap-judgements  (*that 15/17 would still work*) -- but I don't know the hook providers/consumers very well.

So @colemanw I would encourage you to do the same exercise (since you probably do know the hook better; and maybe you've already done this, but I didn't any mention). Should only take a few minutes. 

* Skim the hooks in here. My guess was they looked compatible:

```
    areas/areas.php
    automaticsoftcredit/automaticsoftcredit.php
    biz.lcdservices.civizoom/civizoom.php
    civimoodle/civimoodle.php
    civirepresent/civirepresent.php
    com.aghstrategies.petitionemail/petitionemail.php
    languagecustomfield/languagecustomfield.php
    missingoptionvalues/missingoptionvalues.php
    paymentmethodbydomain/paymentmethodbydomain.php
    synopsis/synopsis.php
    timezone/timezone.php
    uk.co.compucorp.membershipextras/membershipextras.php
    uk.co.mjwconsult.variablerecurpayments/variablerecurpayments.php
    uk.co.vedaconsulting.mautic/mautic.php
    variablerecurpayments/variablerecurpayments.php
    ```

* Skim the hooks in here. My guess is that they have compat issues:

    ```
    org.civicrm.volunteer/volunteer.php (inspects context, but context is changing)
    l10nx/CRM/L10nx/Feeder.php          (inspects both $options and context, which are both changing)
    ```

**Comment by colemanw** [comment]
> @totten changing the format of array $options and also changing the interpretation of $params['context'] sounds like an invitation for some twisty-turny situations. (From a contract POV, it sounds like enough to consider a different hook name...

I considered that, but then thought about:

1. Would the new hook need a different signature?
  - Nah the current signature is fine
2. Would current implementors be less broken if we migrated to a new hook?
  - No they'd be more broken. Unless we added a really inefficient/lossy adapter that flattened the options, called the old hook, and then unflattened them.
3. Is this worse than the status quo?
  - Not by much. The current hook signature is already bonkers, with a crazy `context` param that can literally turn the `$options` inside out and backwards (with most hook implementations not accounting for this possibility so they're already broken in some contexts).
  - Also, think about this: every once in a while a new version has *added* to the list of possible contexts, which quietly throws all hook implementations for another loop, with yet another crazy permutation they're not prepared to handle.
 
And yes, here we are, doing it again and adding yet another `context`, so that's a step back, but IMO we're also taking 3 steps forward:
- From now on, `context` will never be anything but "full" - hook implementations no longer need live in fear of it changing and flipping everything on its head.
- From now on, there will only be one format for the `$options` array passed to the hook. Transformations now happen sensibly *after* the hook is called, not before.
- The new "full" format is, as you saw in `universe` mostly compatible with existing hook implementations, as they usually care about setting or unsetting items by key, and those are unchanged.

**Comment by colemanw** [comment]
@totten reviewing all those hook implementations, the biggest breakage was actually in the custom field naming, which this PR has the effect of changing from `custom_123` to `GroupName.FieldName`.

I went through and submitted PRs/MRs to update all the extensions that needed it, and my takeaway is that while transitions are no fun, this is ultimately a really good change, because in all cases the extension code was contorting to look up the custom field id just so they could match that old `custom_123` format, and the new `GroupName.FieldName` format is much easier for them to work with.

**Comment by totten** [comment]
Oh, wow! Tour de force.

I agree with you - it's a bit of change, but the new contract looks easier to understand, and it's mostly compatible, so we should do it for 5.79. :+1: (*not branched yet; so still `merge ready`*)

**Comment by colemanw** [comment]
Looks like we're branched. Here goes!!

**Pull request merged**